validator/src/main/java/com/google/daq/mqtt/util/MessageUpgrader.java

```diff
@@ -48,30 +48,53 @@ public MessageUpgrader(String schemaName, JsonNode message) {
 
   /**
    * Update message to the latest standard.
+   *
+   * @param forceUpgrade true to force a complete upgrade pass irrespective of original version
+   * @return true if the message has been altered
    */
-  public void upgrade() {
+  public boolean upgrade(boolean forceUpgrade) {
     if (major != 1) {
       throw new IllegalArgumentException("Starting major version " + major);
     }
+
+    final JsonNode original = message.deepCopy();
+    boolean upgraded = false;
+
+    if (forceUpgrade || minor < 0) {
+      minor = 0;
+      patch = 0;
+      upgraded = true;
+    }
+
     if (minor < 3) {
-      upgrade_1_3();
+      minor = 3;
+      patch = 0;
     }
+
     if (minor == 3 && patch < 14) {
+      JsonNode before = message.deepCopy();
       upgrade_1_3_14();
+      upgraded |= !before.equals(message);
+      patch = 14;
     }
-    if (message.has(VERSION_PROPERTY_KEY)) {
+
+    if (minor < 4) {
```

Comment:
I think  the new operational block needs to be in a new version `1.4.1`, and the upgrade is then from 1.4.0 to 1.4.1?